### PR TITLE
[FIX] l10n_de_stock,stock: delivery address at the right place

### PR DESCRIPTION
**Pull request by MerlinGuillaume**
'Delivery Slip' report with the document layout 'DIN5008' needs to have
the delivery address in the hole of the envelope

Steps to reproduce:
1. Install Sales app as well as Drop Shipping and l10n_de module
2. Go to General Settings -> Document Layout and select
external_layout_DIN5008
3. Specify a Dropship route for a product
4. Create and confirm an order for that product
5. Open the purchase order with the 'Purchase' tab
6. Confirm the corresponding purchase order
7. Open the 'Receipt' tab
8. Print the Delivery Slip
9. The document reference is above the address instead of under it

Solution:
Add the `l10n_de_addresses` computed field and set the addresses through
 `<t t-set="address">` and `<t t-set="information_block"/>`

OPW-2709866
OPW-2715851

### CODE REVIEW

**Review by nle-odoo — COMMENTED** [review]
hi

reproducing your use case in 14.0, I don't see the address below the title in the delivery slip:

> 2022-01-26-133927_801x547_scrot

so unless I am missing something, this doesn't seem needed in 14.0

testing in 15.0 I am able to reproduce the addresses below the title:

> 2022-01-26-140840_805x652_scrot

with your change if I update only the stock module I see this:

> 2022-01-26-140623_804x733_scrot

if I update only l10n_de I see this:

> 2022-01-26-141040_801x570_scrot

and if I update both I see this, with both address above title but not in top left:

> 2022-01-26-141210_802x652_scrot

so I'm not sure what would be best, in my opinion:

- maybe the addresses like they currently are already alright.

- or the report should be using the `l10n_de_addresses` field and fallback to `<t t-set="information_block">` instead of what there is currently with the addresses directly added in the report

- or there should something similar to this fix, but with instead of what is currently have a variable such as `<t t-set="do_not_print_title" t-value="True"/>` and a CSS class that allow to override:
 
https://github.com/odoo/odoo/blob/0b43f52116fd9a58fb0a123db576d822fd541043/addons/l10n_de/static/src/scss/report_din5008.scss#L69-L71

@william-andre what do you think? In my opinion best would be to change nothing, or if it made enough sense, do something like what is done in sale.order and purchase.order (with l10n_de_addresses and `<t t-set="information_block"/>`)

**Comment by MerlinGuillaume** [comment]
@nle-odoo The issue in v14 is only with the document 'Picking Operations', I thought the two problems were related and wasn't sure if I was supposed to do 2 PR for the two documents

**Comment by nle-odoo** [comment]
> @nle-odoo The issue in v14 is only with the document 'Picking Operations', I thought the two problems were related and wasn't sure if I was supposed to do 2 PR for the two documents

you can only do one PR to fix a similar issue in two documents (especially if it's the same module)

but from what I understand the two linked tickets only report an issue in 15.0 for "Delivery slip", so it seems to me that fixing another report in 14.0 should not be needed.


Checking further, the "Delivery Slip" change location of address since this change https://github.com/odoo/odoo/commit/af13e76629d0ad1684f2be26a578f210ea4f1552 and that was done so "delivery slip show addresses the same way as picking operations" (but it might not have taken into account din5008 formatting).

So in my opinion we should still do nothing, or in 15.0 add the `l10n_de_addresses` computed field and `<t t-set="information_block"/>` inside the report templates if we think it should be done.

**Review by nle-odoo — COMMENTED** [review]
with current change I think we will have:

- report_deliveryslip (this report) + external_layout_din5008 without XML update => Vendor/Warehouse/Customer in top left, besides that no change (Delivery/Warehouse + Vendor/Warehouse/Customer below title)

- report_deliveryslip (this report) + external_layout_din5008 with XML update => Delivery/Warehouse + Vendor/Warehouse/Customer in top left

- report_deliveryslip (this report) + external_layout_din5008 without XML update =>  nothing change

- report_deliveryslip (this report) + external_layout_din5008 with XML update => addresses above title

- other stock.picking reports + external_layout_din5008 without XML update => Vendor/Warehouse/Customer in top left, besides that nothing change (Delivery/Warehouse + Vendor/Warehouse/Customer below title)

- other stock.picking reports + external_layout_din5008 with XML update => Vendor/Warehouse/Customer added above title, besides that nothing change (Delivery/Warehouse + Vendor/Warehouse/Customer below title)

- other stock.picking reports + other layout without XML update => nothing change

- other stock.picking reports + other layout with XML update => nothing change

tldr; in summary I think there is only a significant change for external_layout_din5008 layout:

=> for report_deliveryslip without updating XML:  you have one address in top-left, two address below title
=> for report_deliveryslip if you update XML: you have two address in top-left, no address below title
=> for other stock.picking reports with or without updating XML: you have one new address in top-left, no other change (so for stock.report_picking no change)

So it looks good for me if it's ok like that (and if the address we could expect is in top-left when there is only one address added in top-left).

If we would not want to modify other stock.picking reports, then I think the only solution would be to only add `<t t-set="address">` and not adding the `_compute_l10n_de_addresses` or `<t t-set="information_block">`.

**Comment by Whenrow** [comment]
I see you already discuss that. It would be better to leave the stock template unchanged and make your correction only in l10n_de_stock. I don't want to break this patch in case we change the delivery slip (I mean, the least possible)

**Comment by MerlinGuillaume** [comment]
I tried to fix the issues mentioned by NLE in this comment https://github.com/odoo/odoo/pull/82208#pullrequestreview-874915485 but I end up breaking something else instead so I'm a bit stuck and don't know what to do to improve on it.
To summarize, here are the changes made by this fix:
#### With a normal order:
- No document layout
**Delivery Slip** : Customer Address moved to the left but Delivery Address added (they use the same partner) :-1: 
**Picking Operations** : No changes
- external_layout_din5008
**Delivery Slip** : Customer Address renamed Delivery Address and moved in the top left
**Picking Operations** : No changes 

#### With a drop-shipping order:
- No document layout
**Delivery Slip** : Customer Address and Vendor Address switched places
**Picking Operations** : No changes
- external_layout_din5008
**Delivery Slip** : Delivery Address moved to the top left with Vendor Address just below it
**Picking Operations** : Vendor Address added on top of the title :-1:

The Delivery Address is always at the correct place when using the DIN5008 layout but the fix modifies other reports with or without this layout.

**Comment by Whenrow** [comment]
robodoo r+
